src/phoenix/trace/openai/instrumentor.py

```diff
@@ -0,0 +1,279 @@
+import json
+from datetime import datetime
+from enum import Enum
+from inspect import signature
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Dict,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Union,
+    cast,
+)
+
+from phoenix.trace.schemas import (
+    AttributeValue,
+    Message,
+    SpanAttributes,
+    SpanEvent,
+    SpanException,
+    SpanKind,
+    SpanStatusCode,
+)
+from phoenix.trace.semantic_conventions import (
+    INPUT_MIME_TYPE,
+    INPUT_VALUE,
+    LLM_FUNCTION_CALL,
+    LLM_INPUT_MESSAGES,
+    LLM_INVOCATION_PARAMETERS,
+    LLM_TOKEN_COUNT_COMPLETION,
+    LLM_TOKEN_COUNT_PROMPT,
+    LLM_TOKEN_COUNT_TOTAL,
+    MESSAGE_CONTENT,
+    MESSAGE_FUNCTION_CALL_ARGUMENTS_JSON,
+    MESSAGE_FUNCTION_CALL_NAME,
+    MESSAGE_NAME,
+    MESSAGE_ROLE,
+    OUTPUT_MIME_TYPE,
+    OUTPUT_VALUE,
+    MimeType,
+)
+from phoenix.trace.utils import get_stacktrace, import_package
+
+from ..tracer import Tracer
+
+if TYPE_CHECKING:
+    from openai.openai_response import OpenAIResponse
+
+
+Parameters = Mapping[str, Any]
+
+INSTRUMENTED_ATTRIBUTE_NAME = "is_instrumented_with_openinference_tracer"
+
+
+class NotSet:
+    """Indicates that no value is to be set for a given attribute."""
```

Comment:
What do we gain from this new approach compared to just leaving the keys out? The downside of this new object is that it would complicate the code paths for JSON and protobuf. It also seems redundant: isn't "not set" the same as leaving the keys out?

Reply:
`NotSet` is just a sentinel that tells us to leave the keys out, we're never passing that object over the wire

Reply:
Because this object is not directly serializable, wouldn't all of our serialization code paths need to know about this object so they can skip serializing it? What is the benefit of keeping a sentinel that tells us to leave the keys out versus directly leaving the keys out? And isn't the latter simpler?

Reply:
This object is always skipped! The sentinel is used by ` _to_span_attributes` to determine whether or not to include the key. If the value is `NotSet` then no attribute is attached to the span.

Reply:
@RogerHYang we're advocating for exactly the same behavior. The role of `NotSet` here was just to make the "exclusion" behavior explicit in our code as opposed to relying on the implicit behavior of generators yielding nothing. Spelling out what we want with the `NotSet` mechanism I think is a little bit less prone to being an issue in the future

Reply:
This seems like a roundabout way to arrive at the same goal and without a compelling objective for the detour. I don't want to sound like I am overly attached to my original approach or anything 😅 but we try to keep things as lean as possible because minimalism also has its benefits. I am not disputing that this mechanism has real benefits, but is there a pressing need for its addition?

Reply:
> being an issue in the future

PS. our operating principle is not deal with the future until the future.

Reply:
resolved offline